File: challenge.rb
require 'time'
require 'concurrent'

class Station
  attr_reader :min, :sum, :max, :count

  def initialize
    @min = Float::INFINITY
    @sum = 0
    @max = -Float::INFINITY
    @count = 0
    @mutex = Mutex.new
  end

  def add_temperature(temp)
    @mutex.synchronize do
      @min = temp if temp < @min
      @max = temp if temp > @max
      @sum += temp
      @count += 1
    end
  end

  def average_temperature
    @count == 0 ? 0 : @sum / @count
  end
end

class ProgressTracker
  def initialize(total_lines)
    @total_lines = total_lines.to_f
    @processed_lines_count = Concurrent::AtomicFixnum.new(0)
  end

  def increment_processed_lines(count = 1)
    @processed_lines_count.update { |value| value + count }
  end

  def print_progress
    current_percentage = (@processed_lines_count.value / @total_lines) * 100
    print "\rProcessing... #{current_percentage.round(2)}% complete"
  end

  def finalize_progress
    puts "\nProcessing completed."
  end
end

stations = Concurrent::Hash.new { |hash, key| hash[key] = Station.new }
total_lines = 1_000_000_000 # Predefined total lines in the file
progress_tracker = ProgressTracker.new(total_lines)

def parse_line(line)
  parts = line.split(';')
  station = parts[0]
  temperature = parts[1].to_f
  [station, temperature]
end

def process_subchunk(subchunk, stations)
  subchunk.each do |line|
    station, temperature = parse_line(line)
    stations[station].add_temperature(temperature)
  end
end

def process_chunk(chunk, stations, subchunk_size, progress_tracker)
  futures = []
  
  chunk.each_slice(subchunk_size) do |subchunk|
    futures << Concurrent::Future.execute { process_subchunk(subchunk, stations) }
  end
  
  futures.each(&:value) # Wait for all futures to complete
  progress_tracker.increment_processed_lines(chunk.size)
  progress_tracker.print_progress
end

def process_data(filename, stations, buffer_size, subchunk_size, total_lines, progress_tracker)
  start_time = Time.now

  pool = Concurrent::FixedThreadPool.new(8) # Adjust the number of threads as needed

  buffer = []

  File.foreach(filename) do |line|
    buffer << line

    if buffer.length >= buffer_size
      chunk_to_process = buffer.shift(buffer_size) # Remove processed lines from buffer immediately
      pool.post { process_chunk(chunk_to_process, stations, subchunk_size, progress_tracker) } # Process the chunk in parallel
    end
  end

  # Process any remaining lines in the buffer
  unless buffer.empty?
    chunk_to_process = buffer.shift(buffer.size) # Remove processed lines from buffer immediately
    pool.post { process_chunk(chunk_to_process, stations, subchunk_size, progress_tracker) } # Process the chunk in parallel
  end

  pool.shutdown
  pool.wait_for_termination

  end_time = Time.now
  puts "\nTotal processing time: #{end_time - start_time} seconds"
  progress_tracker.finalize_progress
  print_data(stations)
end

def print_data(stations)
  sorted_stations = stations.keys.sort
  print '{'
  sorted_stations.each_with_index do |station, index|
    stats = stations[station]
    avg_temp = stats.average_temperature
    print "#{station}=#{format('%.1f', stats.min)}/#{format('%.1f', avg_temp)}/#{format('%.1f', stats.max)}"
    print ',' if index < sorted_stations.length - 1
  end
  print "}\n"
end

# Replace 'measurements.txt' with your actual filename
buffer_size = 10 * (4 * 250000) # Adjust the buffer size as needed
subchunk_size = buffer_size / 8 # Size of each subchunk
process_data('measurements.txt', stations, buffer_size, subchunk_size, total_lines, progress_tracker)

File: challenge.py
import os
import json
from multiprocessing import Pool, cpu_count

class Station:
    def __init__(self):
        self.min = float('inf')
        self.sum = 0
        self.max = float('-inf')
        self.count = 0

    def add_temperature(self, temp):
        if temp < self.min:
            self.min = temp
        if temp > self.max:
            self.max = temp
        self.sum += temp
        self.count += 1

    def average_temperature(self):
        if self.count == 0:
            return 0
        return self.sum / self.count

def process_chunk(chunk):
    stations = {}
    for line in chunk:
        if line.strip() == '':
            continue
        
        try:
            station, temperature = line.split(';')
            temperature = float(temperature)
        except ValueError:
            continue
        
        if station not in stations:
            stations[station] = Station()
        stations[station].add_temperature(temperature)
    
    encoded_lines = []
    for station, station_obj in stations.items():
        data_string = json.dumps({'station': station, 'temperature': station_obj.sum / station_obj.count}) + '\n'
        encoded_lines.append(data_string.encode('utf-8'))
    
    return encoded_lines

def process_output(encoded_chunks):
    stations = {}
    for encoded_lines in encoded_chunks:
        for encoded_line in encoded_lines:
            json_string = encoded_line.decode('utf-8')
            try:
                data = json.loads(json_string)
                station, temperature = data['station'], data['temperature']
            except json.JSONDecodeError:
                continue
            
            if station not in stations:
                stations[station] = Station()
            stations[station].add_temperature(temperature)
    
    return stations

def process_data(filename, chunk_size=4*1024*1024):
    pool = Pool(cpu_count())
    file_size = os.path.getsize(filename)
    total_chunks = (file_size + chunk_size - 1) // chunk_size
    progress_bar_length = 50
    
    def progress_bar(progress):
        percentage = min(100, round(progress * 100))
        progress_length = min(progress_bar_length, round(progress_bar_length * progress))
        bar = '=' * progress_length + '-' * (progress_bar_length - progress_length)
        print(f'\r[{bar}] {percentage}%', end='')

    with open(filename, 'r') as file:
        encoded_chunks = []
        partial_line = ''
        
        for chunk_index in range(total_chunks):
            chunk = file.read(chunk_size)
            if not chunk:
                break
            
            lines = (partial_line + chunk).split('\n')
            partial_line = lines.pop()  # Save the incomplete line fragment
            
            encoded_chunks.append(pool.apply_async(process_chunk, (lines,)))
            
            progress_bar((chunk_index + 1) / total_chunks)
        
        if partial_line:
            encoded_chunks.append(pool.apply_async(process_chunk, ([partial_line],)))
        
        pool.close()
        pool.join()
        
        encoded_chunks = [result.get() for result in encoded_chunks]
    
    stations = process_output(encoded_chunks)
    return stations

def print_data(stations):
    sorted_stations = sorted(stations.keys())
    result = '{'
    
    for index, station in enumerate(sorted_stations):
        stats = stations[station]
        avg_temp = stats.average_temperature()
        result += f'{station}={stats.min:.1f}/{avg_temp:.1f}/{stats.max:.1f}'
        if index < len(sorted_stations) - 1:
            result += ','
    
    result += '}'
    print(result)

if __name__ == '__main__':
    import time
    start_time = time.time()
    
    stations = process_data('measurements.txt')
    
    print(f'\nProcessing completed in {time.time() - start_time:.2f} seconds.')
    print_data(stations)
